# Preparing Chicago Weather Dataset

We want to examine whether the weather has an effect on the type of crimes carried out in the time surrounding sports games. For example, if the weather is bad are more fans staying indoors to watch the games and in turn is domestic violence rising? Then alternatively to that, does good weather cause an increase in assault and vandalism, as more fans would be outdoor and attending the games?

In this notebook we will load in three datasets: one for temperature, one for weather description and one for wind speed. Each one contains data on 36 cities in the US and their columns is named after each one. We want to grade this weather so we can get an overall sense of how good or bad the weather was at certain times in the day. Each type of weather will contribute equally to determine what type of weather occurred. 

### Loading in the data

In [1020]:
import os.path
import datetime
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline

The temperature dataset is loaded in. This contains hourly data of 36 different cities in the US. The datetime column is made the index column. 

In [1021]:
if not os.path.exists( "../../data/raw/temperature.csv" ):
    print("Missing dataset file")

In [1022]:
temp = pd.read_csv( "../../data/raw/temperature.csv", index_col="datetime", parse_dates=True)

In [1023]:
temp.head()

Vancouver    Portland  San Francisco     Seattle  \
datetime                                                                 
2012-10-01 12:00:00         NaN         NaN            NaN         NaN   
2012-10-01 13:00:00  284.630000  282.080000     289.480000  281.800000   
2012-10-01 14:00:00  284.629041  282.083252     289.474993  281.797217   
2012-10-01 15:00:00  284.626998  282.091866     289.460618  281.789833   
2012-10-01 16:00:00  284.624955  282.100481     289.446243  281.782449   

                     Los Angeles   San Diego   Las Vegas     Phoenix  \
datetime                                                               
2012-10-01 12:00:00          NaN         NaN         NaN         NaN   
2012-10-01 13:00:00   291.870000  291.530000  293.410000  296.600000   
2012-10-01 14:00:00   291.868186  291.533501  293.403141  296.608509   
2012-10-01 15:00:00   291.862844  291.543355  293.392177  296.631487   
2012-10-01 16:00:00   291.857503  291.553209  293.381213  296.654466   

                     Albuquerque      Denver    ...      Philadelphia  \
datetime                                        ...                     
2012-10-01 12:00:00          NaN         NaN    ...               NaN   
2012-10-01 13:00:00   285.120000  284.610000    ...        285.630000   
2012-10-01 14:00:00   285.154558  284.607306    ...        285.663208   
2012-10-01 15:00:00   285.233952  284.599918    ...        285.756824   
2012-10-01 16:00:00   285.313345  284.592530    ...        285.850440   

                       New York    Montreal      Boston   Beersheba  \
datetime                                                              
2012-10-01 12:00:00         NaN         NaN         NaN         NaN   
2012-10-01 13:00:00  288.220000  285.830000  287.170000  307.590000   
2012-10-01 14:00:00  288.247676  285.834650  287.186092  307.590000   
2012-10-01 15:00:00  288.326940  285.847790  287.231672  307.391513   
2012-10-01 16:00:00  288.406203  285.860929  287.277251  307.145200   

                     Tel Aviv District       Eilat  Haifa  Nahariyya  \
datetime                                                               
2012-10-01 12:00:00                NaN  309.100000    NaN        NaN   
2012-10-01 13:00:00         305.470000  310.580000  304.4      304.4   
2012-10-01 14:00:00         304.310000  310.495769  304.4      304.4   
2012-10-01 15:00:00         304.281841  310.411538  304.4      304.4   
2012-10-01 16:00:00         304.238015  310.327308  304.4      304.4   

                     Jerusalem  
datetime                        
2012-10-01 12:00:00        NaN  
2012-10-01 13:00:00      303.5  
2012-10-01 14:00:00      303.5  
2012-10-01 15:00:00      303.5  
2012-10-01 16:00:00      303.5  

[5 rows x 36 columns]

### Chicago

Filtering the column from Chicago...

In [1024]:
chi1 = temp.filter(items=['Chicago'])

In [1025]:
chi1.head()

Chicago
datetime                       
2012-10-01 12:00:00         NaN
2012-10-01 13:00:00  284.010000
2012-10-01 14:00:00  284.054691
2012-10-01 15:00:00  284.177412
2012-10-01 16:00:00  284.300133

This column is renamed to 'Temperature'. In each dataset, whether it be the temperature, wind speed or description dataset, the column is named after the city. When we would join them, for this example we would be left with three columns named Chicago. 

In [1026]:
chi1 = chi1.rename(columns={'Chicago': 'Temperature'})

In [1027]:
chi1.head()

Temperature
datetime                        
2012-10-01 12:00:00          NaN
2012-10-01 13:00:00   284.010000
2012-10-01 14:00:00   284.054691
2012-10-01 15:00:00   284.177412
2012-10-01 16:00:00   284.300133

The wind speed dataset is now loaded in...

In [1028]:
if not os.path.exists( "../../data/raw/wind_speed.csv" ):
    print("Missing dataset file")

The datetime column is again made the index..

In [1029]:
temp1 = pd.read_csv( "../../data/raw/wind_speed.csv", index_col="datetime", parse_dates=True)

In [1030]:
temp1.head()

Vancouver  Portland  San Francisco  Seattle  Los Angeles  \
datetime                                                                        
2012-10-01 12:00:00        NaN       NaN            NaN      NaN          NaN   
2012-10-01 13:00:00        0.0       0.0            2.0      0.0          0.0   
2012-10-01 14:00:00        0.0       0.0            2.0      0.0          0.0   
2012-10-01 15:00:00        0.0       0.0            2.0      0.0          0.0   
2012-10-01 16:00:00        0.0       0.0            2.0      0.0          0.0   

                     San Diego  Las Vegas  Phoenix  Albuquerque  Denver  \
datetime                                                                  
2012-10-01 12:00:00        NaN        NaN      NaN          NaN     NaN   
2012-10-01 13:00:00        0.0        0.0      2.0          4.0     4.0   
2012-10-01 14:00:00        0.0        0.0      2.0          4.0     4.0   
2012-10-01 15:00:00        0.0        0.0      2.0          4.0     3.0   
2012-10-01 16:00:00        0.0        0.0      2.0          4.0     3.0   

                       ...      Philadelphia  New York  Montreal  Boston  \
datetime               ...                                                 
2012-10-01 12:00:00    ...               NaN       NaN       NaN     NaN   
2012-10-01 13:00:00    ...               4.0       7.0       4.0     3.0   
2012-10-01 14:00:00    ...               4.0       7.0       4.0     3.0   
2012-10-01 15:00:00    ...               3.0       7.0       4.0     3.0   
2012-10-01 16:00:00    ...               3.0       7.0       4.0     3.0   

                     Beersheba  Tel Aviv District  Eilat  Haifa  Nahariyya  \
datetime                                                                     
2012-10-01 12:00:00        NaN                NaN    8.0    NaN        NaN   
2012-10-01 13:00:00        1.0                0.0    8.0    2.0        2.0   
2012-10-01 14:00:00        3.0                0.0    8.0    2.0        2.0   
2012-10-01 15:00:00        3.0                0.0    8.0    2.0        2.0   
2012-10-01 16:00:00        3.0                0.0    8.0    2.0        2.0   

                     Jerusalem  
datetime                        
2012-10-01 12:00:00        NaN  
2012-10-01 13:00:00        2.0  
2012-10-01 14:00:00        2.0  
2012-10-01 15:00:00        2.0  
2012-10-01 16:00:00        2.0  

[5 rows x 36 columns]

Again only interested in the Chicago column

In [1031]:
chi2 = temp1.filter(items=['Chicago'])

In [1032]:
chi2.head()

Chicago
datetime                    
2012-10-01 12:00:00      NaN
2012-10-01 13:00:00      0.0
2012-10-01 14:00:00      0.0
2012-10-01 15:00:00      0.0
2012-10-01 16:00:00      0.0

The Chicago column is renamed to 'Wind Speed'

In [1033]:
chi2 = chi2.rename(columns={'Chicago': 'Wind Speed'})

In [1034]:
chi2.head()

Wind Speed
datetime                       
2012-10-01 12:00:00         NaN
2012-10-01 13:00:00         0.0
2012-10-01 14:00:00         0.0
2012-10-01 15:00:00         0.0
2012-10-01 16:00:00         0.0

The weather description dataset is loaded in..

In [1035]:
if not os.path.exists( "../../data/raw/weather_description.csv" ):
    print("Missing dataset file")

datetime is again made the index..

In [1036]:
temp2 = pd.read_csv( "../../data/raw/weather_description.csv", index_col="datetime", parse_dates=True)

In [1037]:
temp2.head()

Vancouver          Portland San Francisco  \
datetime                                                             
2012-10-01 12:00:00            NaN               NaN           NaN   
2012-10-01 13:00:00           mist  scattered clouds    light rain   
2012-10-01 14:00:00  broken clouds  scattered clouds  sky is clear   
2012-10-01 15:00:00  broken clouds  scattered clouds  sky is clear   
2012-10-01 16:00:00  broken clouds  scattered clouds  sky is clear   

                          Seattle   Los Angeles     San Diego     Las Vegas  \
datetime                                                                      
2012-10-01 12:00:00           NaN           NaN           NaN           NaN   
2012-10-01 13:00:00  sky is clear          mist  sky is clear  sky is clear   
2012-10-01 14:00:00  sky is clear  sky is clear  sky is clear  sky is clear   
2012-10-01 15:00:00  sky is clear  sky is clear  sky is clear  sky is clear   
2012-10-01 16:00:00  sky is clear  sky is clear  sky is clear  sky is clear   

                          Phoenix   Albuquerque         Denver  \
datetime                                                         
2012-10-01 12:00:00           NaN           NaN            NaN   
2012-10-01 13:00:00  sky is clear  sky is clear     light rain   
2012-10-01 14:00:00  sky is clear  sky is clear  broken clouds   
2012-10-01 15:00:00  sky is clear  sky is clear  broken clouds   
2012-10-01 16:00:00  sky is clear  sky is clear  broken clouds   

                          ...          Philadelphia    New York  \
datetime                  ...                                     
2012-10-01 12:00:00       ...                   NaN         NaN   
2012-10-01 13:00:00       ...         broken clouds  few clouds   
2012-10-01 14:00:00       ...         broken clouds  few clouds   
2012-10-01 15:00:00       ...         broken clouds  few clouds   
2012-10-01 16:00:00       ...         broken clouds  few clouds   

                            Montreal        Boston        Beersheba  \
datetime                                                              
2012-10-01 12:00:00              NaN           NaN              NaN   
2012-10-01 13:00:00  overcast clouds  sky is clear     sky is clear   
2012-10-01 14:00:00     sky is clear    few clouds     sky is clear   
2012-10-01 15:00:00     sky is clear    few clouds  overcast clouds   
2012-10-01 16:00:00     sky is clear    few clouds  overcast clouds   

                    Tel Aviv District          Eilat            Haifa  \
datetime                                                                
2012-10-01 12:00:00               NaN           haze              NaN   
2012-10-01 13:00:00      sky is clear           haze     sky is clear   
2012-10-01 14:00:00      sky is clear  broken clouds  overcast clouds   
2012-10-01 15:00:00      sky is clear  broken clouds  overcast clouds   
2012-10-01 16:00:00      sky is clear  broken clouds  overcast clouds   

                           Nahariyya        Jerusalem  
datetime                                               
2012-10-01 12:00:00              NaN              NaN  
2012-10-01 13:00:00     sky is clear     sky is clear  
2012-10-01 14:00:00     sky is clear  overcast clouds  
2012-10-01 15:00:00  overcast clouds  overcast clouds  
2012-10-01 16:00:00  overcast clouds  overcast clouds  

[5 rows x 36 columns]

Again, only looking at the Chicago column

In [1038]:
chi3 = temp2.filter(items=['Chicago'])

In [1039]:
chi3.head()

Chicago
datetime                            
2012-10-01 12:00:00              NaN
2012-10-01 13:00:00  overcast clouds
2012-10-01 14:00:00  overcast clouds
2012-10-01 15:00:00  overcast clouds
2012-10-01 16:00:00  overcast clouds

This column is renamed to 'Description'

In [1040]:
chi3 = chi3.rename(columns={'Chicago': 'Description'})

In [1041]:
chi3.head()

Description
datetime                            
2012-10-01 12:00:00              NaN
2012-10-01 13:00:00  overcast clouds
2012-10-01 14:00:00  overcast clouds
2012-10-01 15:00:00  overcast clouds
2012-10-01 16:00:00  overcast clouds

The Temperature, Wind Speed and Description columns are now concatted into the one dataframe. 

In [1042]:
frames = [chi1, chi2, chi3]
df = pd.concat(frames, axis = 1)

In [1043]:
df.head()

Temperature  Wind Speed      Description
datetime                                                     
2012-10-01 12:00:00          NaN         NaN              NaN
2012-10-01 13:00:00   284.010000         0.0  overcast clouds
2012-10-01 14:00:00   284.054691         0.0  overcast clouds
2012-10-01 15:00:00   284.177412         0.0  overcast clouds
2012-10-01 16:00:00   284.300133         0.0  overcast clouds

### Dealing with Null values

We look at the null values in the new dataframe. 

In [1044]:
df.isnull().sum()

Temperature    3
Wind Speed     1
Description    1
dtype: int64

Looking at the rows where Description is null... The only example is in the first row which is black in all fields so this row is deleted. 

In [1045]:
print(df[df["Description"].isnull()])

                     Temperature  Wind Speed Description
datetime                                                
2012-10-01 12:00:00          NaN         NaN         NaN


In [1046]:
df = df.dropna(subset = ['Description'])

Looking at the temperature fields that are null. 

In [1047]:
print(df[df["Temperature"].isnull()])

                     Temperature  Wind Speed Description
datetime                                                
2013-03-11 07:00:00          NaN         1.0  light rain
2013-03-11 08:00:00          NaN         2.0  light rain


Looking at the day that consists of the null values. 

In [1048]:
testdate = df.loc['2013-03-11':'2013-03-11']

In [1049]:
testdate

Temperature  Wind Speed      Description
datetime                                                     
2013-03-11 00:00:00      278.920         5.0  overcast clouds
2013-03-11 01:00:00      279.005         4.0  overcast clouds
2013-03-11 02:00:00      279.090         4.0  overcast clouds
2013-03-11 03:00:00      279.380         3.0  overcast clouds
2013-03-11 04:00:00      280.220         3.0  overcast clouds
2013-03-11 05:00:00      280.380         4.0  overcast clouds
2013-03-11 06:00:00      280.690         3.0    broken clouds
2013-03-11 07:00:00          NaN         1.0       light rain
2013-03-11 08:00:00          NaN         2.0       light rain
2013-03-11 09:00:00      278.850         4.0             mist
2013-03-11 10:00:00      279.070         5.0       light rain
2013-03-11 11:00:00      279.480         4.0             mist
2013-03-11 12:00:00      279.080         3.0    moderate rain
2013-03-11 13:00:00      279.265         3.0  overcast clouds
2013-03-11 14:00:00      279.450         4.0             mist
2013-03-11 15:00:00      280.040         4.0             mist
2013-03-11 16:00:00      279.053         3.0  overcast clouds
2013-03-11 17:00:00      281.350         3.0             mist
2013-03-11 18:00:00      281.750         2.0     sky is clear
2013-03-11 19:00:00      282.150         1.0  overcast clouds
2013-03-11 20:00:00      282.410         1.0             mist
2013-03-11 21:00:00      281.900         0.0             mist
2013-03-11 22:00:00      278.976         4.0       light rain
2013-03-11 23:00:00      281.440         4.0              fog

Find the average between 280.690 and 278.850. This is equal to 279.770. We'll then fill the null values with this average. Instead of deleting the rows we felt this was the best solution as we could need these rows later when comparing them to the NFL and NBA games. We felt getting the average between the row after and before the missing values would give us the best result. 

In [1050]:
df['Temperature'] = df['Temperature'].fillna(279.770)

In [1051]:
print(df[df["Temperature"].isnull()])

Empty DataFrame
Columns: [Temperature, Wind Speed, Description]
Index: []


We want to convert the temperature to degrees Celcius from kelvin. Celcius is more familiar to us and the general public so we felt it would make understanding easier. The formula to do this is T(°C)=T(K)-273.15, Therefore, we create a new column called 'Celcius' and use this formula. 

In [1052]:
df["Celcius"] = df["Temperature"] - 273.15

In [1053]:
df.head()

Temperature  Wind Speed      Description    Celcius
datetime                                                                
2012-10-01 13:00:00   284.010000         0.0  overcast clouds  10.860000
2012-10-01 14:00:00   284.054691         0.0  overcast clouds  10.904691
2012-10-01 15:00:00   284.177412         0.0  overcast clouds  11.027412
2012-10-01 16:00:00   284.300133         0.0  overcast clouds  11.150133
2012-10-01 17:00:00   284.422855         0.0  overcast clouds  11.272855

The Kelvin temperature column is now dropped as it is no longer needed

In [1054]:
df = df.drop(df.columns[[0]], axis=1)

In [1055]:
chiWea = df

In [1056]:
chiWea.head()

Wind Speed      Description    Celcius
datetime                                                   
2012-10-01 13:00:00         0.0  overcast clouds  10.860000
2012-10-01 14:00:00         0.0  overcast clouds  10.904691
2012-10-01 15:00:00         0.0  overcast clouds  11.027412
2012-10-01 16:00:00         0.0  overcast clouds  11.150133
2012-10-01 17:00:00         0.0  overcast clouds  11.272855

### The Weather Grading System

We now look at all the different types of Weather in terms of description. These will be grouped into different groups depending how good the weather is. 

In [1057]:
pt = chiWea.groupby('Description')[('Celcius')].count()
pt.sort_values(ascending = False)

Description
sky is clear                           10844
broken clouds                           7255
overcast clouds                         5412
scattered clouds                        4522
mist                                    4112
few clouds                              3851
light rain                              3745
moderate rain                           1469
light snow                               886
haze                                     877
fog                                      530
heavy intensity rain                     503
snow                                     241
light intensity drizzle                  228
heavy snow                               196
proximity thunderstorm                   178
drizzle                                  143
thunderstorm                              78
thunderstorm with light rain              60
very heavy rain                           35
thunderstorm with rain                    27
thunderstorm with heavy rain              2

We create the following function. If the description is 'sky is clear' or 'few clouds' this weather is considered good and is given a value of 3. Then if the description is 'overcast clouds', 'broken clouds' or 'scattered clouds' the weather is considered to be moderate and is given a value of 2. The rest (the bad weather) is given a value of 0. We wanted to get a clear difference between good and even moderate weather which is why bad weather is valued at 0 instead of 1. 

In [1058]:
conditions = [
    (chiWea['Description'] == 'sky is clear') | (chiWea['Description'] == 'few clouds'),
    (chiWea['Description'] == 'overcast clouds') | (chiWea['Description'] == 'broken clouds') 
    | (chiWea['Description'] == 'scattered clouds') ]
choices = ['3', '2']
chiWea['DescRate'] = np.select(conditions, choices, default='0')
print(chiWea)

                     Wind Speed       Description    Celcius DescRate
datetime                                                             
2012-10-01 13:00:00         0.0   overcast clouds  10.860000        2
2012-10-01 14:00:00         0.0   overcast clouds  10.904691        2
2012-10-01 15:00:00         0.0   overcast clouds  11.027412        2
2012-10-01 16:00:00         0.0   overcast clouds  11.150133        2
2012-10-01 17:00:00         0.0   overcast clouds  11.272855        2
2012-10-01 18:00:00         0.0   overcast clouds  11.395576        2
2012-10-01 19:00:00         0.0   overcast clouds  11.518297        2
2012-10-01 20:00:00         0.0   overcast clouds  11.641018        2
2012-10-01 21:00:00         0.0   overcast clouds  11.763739        2
2012-10-01 22:00:00         0.0   overcast clouds  11.886461        2
2012-10-01 23:00:00         0.0   overcast clouds  12.009182        2
2012-10-02 00:00:00         0.0   overcast clouds  12.131903        2
2012-10-02 01:00:00 

The next thing was to determine how good the temperatue was, so we could assign it a grade. We first got the mean temperature. Anything below the average temperature would be considered to be 'bad weather'. 

In [1059]:
chiWea['Celcius'].mean()

10.200414543717171

Again the same type of code is used as above. This time if the temperature is below or equal to 10 a value of 0 is assigned as it is considered bad weather. Anything above or equal to 20 degrees celcius is given a value of 3 as it is good weather. The In between temperatures are considered moderate and is given a value of 2. 

In [1060]:
conditions = [
    (chiWea['Celcius'] <= 10),
    (chiWea['Celcius'] >= 20)]
choices = ['0', '3']
chiWea['TempRate'] = np.select(conditions, choices, default='2')
print(chiWea)

                     Wind Speed       Description    Celcius DescRate TempRate
datetime                                                                      
2012-10-01 13:00:00         0.0   overcast clouds  10.860000        2        2
2012-10-01 14:00:00         0.0   overcast clouds  10.904691        2        2
2012-10-01 15:00:00         0.0   overcast clouds  11.027412        2        2
2012-10-01 16:00:00         0.0   overcast clouds  11.150133        2        2
2012-10-01 17:00:00         0.0   overcast clouds  11.272855        2        2
2012-10-01 18:00:00         0.0   overcast clouds  11.395576        2        2
2012-10-01 19:00:00         0.0   overcast clouds  11.518297        2        2
2012-10-01 20:00:00         0.0   overcast clouds  11.641018        2        2
2012-10-01 21:00:00         0.0   overcast clouds  11.763739        2        2
2012-10-01 22:00:00         0.0   overcast clouds  11.886461        2        2
2012-10-01 23:00:00         0.0   overcast clouds  1

In [1061]:
chiWea.head()

Wind Speed      Description    Celcius DescRate TempRate
datetime                                                                     
2012-10-01 13:00:00         0.0  overcast clouds  10.860000        2        2
2012-10-01 14:00:00         0.0  overcast clouds  10.904691        2        2
2012-10-01 15:00:00         0.0  overcast clouds  11.027412        2        2
2012-10-01 16:00:00         0.0  overcast clouds  11.150133        2        2
2012-10-01 17:00:00         0.0  overcast clouds  11.272855        2        2

The average wind speed is found. This is relatively low. Although there is no indication to what the wind speed is measured in we are inclined to think the measurement is knots. The average wind speed in Chicago is around 6-10km per hour according to https://www.isws.illinois.edu/statecli/wind/wind.htm . This average in knots would fit into this bracket. 

In [1062]:
chiWea['Wind Speed'].mean()

3.7593255546716167

Wind speed less than or equal to 2 is considered to be very calm and is considered good. It is valued at 3. Between 2 and 5 is considered moderate and is valued at 2. Above 5 knots is considered bad and is given a value of 0. 

In [1063]:
conditions = [
    (chiWea['Wind Speed'] <= 2),
    (chiWea['Wind Speed'] > 2) | (chiWea['Wind Speed'] < 5) ]
choices = ['3', '2']
chiWea['WindRate'] = np.select(conditions, choices, default='0')
print(chiWea)

                     Wind Speed       Description    Celcius DescRate  \
datetime                                                                
2012-10-01 13:00:00         0.0   overcast clouds  10.860000        2   
2012-10-01 14:00:00         0.0   overcast clouds  10.904691        2   
2012-10-01 15:00:00         0.0   overcast clouds  11.027412        2   
2012-10-01 16:00:00         0.0   overcast clouds  11.150133        2   
2012-10-01 17:00:00         0.0   overcast clouds  11.272855        2   
2012-10-01 18:00:00         0.0   overcast clouds  11.395576        2   
2012-10-01 19:00:00         0.0   overcast clouds  11.518297        2   
2012-10-01 20:00:00         0.0   overcast clouds  11.641018        2   
2012-10-01 21:00:00         0.0   overcast clouds  11.763739        2   
2012-10-01 22:00:00         0.0   overcast clouds  11.886461        2   
2012-10-01 23:00:00         0.0   overcast clouds  12.009182        2   
2012-10-02 00:00:00         0.0   overcast clouds  

In [1064]:
chiWea.head()

Wind Speed      Description    Celcius DescRate TempRate  \
datetime                                                                        
2012-10-01 13:00:00         0.0  overcast clouds  10.860000        2        2   
2012-10-01 14:00:00         0.0  overcast clouds  10.904691        2        2   
2012-10-01 15:00:00         0.0  overcast clouds  11.027412        2        2   
2012-10-01 16:00:00         0.0  overcast clouds  11.150133        2        2   
2012-10-01 17:00:00         0.0  overcast clouds  11.272855        2        2   

                    WindRate  
datetime                      
2012-10-01 13:00:00        3  
2012-10-01 14:00:00        3  
2012-10-01 15:00:00        3  
2012-10-01 16:00:00        3  
2012-10-01 17:00:00        3

In [1065]:
chiWea.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 45252 entries, 2012-10-01 13:00:00 to 2017-11-30 00:00:00
Freq: H
Data columns (total 6 columns):
Wind Speed     45252 non-null float64
Description    45252 non-null object
Celcius        45252 non-null float64
DescRate       45252 non-null object
TempRate       45252 non-null object
WindRate       45252 non-null object
dtypes: float64(2), object(4)
memory usage: 3.7+ MB


We have to change our new columns to int as we are going to add them together to get our overall weather rating. 

In [1066]:
chiWea['DescRate'] = chiWea['DescRate'].astype(int)

In [1067]:
chiWea['TempRate'] = chiWea['TempRate'].astype(int)

In [1068]:
chiWea['WindRate'] = chiWea['WindRate'].astype(int)

In [1069]:
chiWea.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 45252 entries, 2012-10-01 13:00:00 to 2017-11-30 00:00:00
Freq: H
Data columns (total 6 columns):
Wind Speed     45252 non-null float64
Description    45252 non-null object
Celcius        45252 non-null float64
DescRate       45252 non-null int32
TempRate       45252 non-null int32
WindRate       45252 non-null int32
dtypes: float64(2), int32(3), object(1)
memory usage: 3.1+ MB


The Overall column is created it consits of the Description Rating, Temperature Rating and Wind Rating added together. This overall rating gives us a sense of how good the weather was for that hour. 

In [1070]:
chiWea['Overall'] = chiWea["DescRate"] + chiWea["TempRate"] + chiWea["WindRate"]

In [1071]:
chiWea.head()

Wind Speed      Description    Celcius  DescRate  \
datetime                                                                
2012-10-01 13:00:00         0.0  overcast clouds  10.860000         2   
2012-10-01 14:00:00         0.0  overcast clouds  10.904691         2   
2012-10-01 15:00:00         0.0  overcast clouds  11.027412         2   
2012-10-01 16:00:00         0.0  overcast clouds  11.150133         2   
2012-10-01 17:00:00         0.0  overcast clouds  11.272855         2   

                     TempRate  WindRate  Overall  
datetime                                          
2012-10-01 13:00:00         2         3        7  
2012-10-01 14:00:00         2         3        7  
2012-10-01 15:00:00         2         3        7  
2012-10-01 16:00:00         2         3        7  
2012-10-01 17:00:00         2         3        7

We then want to group the weather into Good, Moderate and Overall Weather. 8 and over is Good. 6 to 8 is considered moderate. And under 6 would be considered to be Bad. For example if we have a case where the temperature is 21 degrees, overcast clouds and rain, this would be considered bad as Raining is always considered bad in terms of weather, and prevents people from going outside. 

In [1072]:
conditions = [
    (chiWea['Overall'] >= 8), 
    (chiWea['Overall'] >= 6) & (chiWea['Overall'] < 8) ]
choices = ['Good', 'Moderate']
chiWea['Weather'] = np.select(conditions, choices, default='Bad')

In [1073]:
chiWea.head(50)

Wind Speed       Description    Celcius  DescRate  \
datetime                                                                 
2012-10-01 13:00:00         0.0   overcast clouds  10.860000         2   
2012-10-01 14:00:00         0.0   overcast clouds  10.904691         2   
2012-10-01 15:00:00         0.0   overcast clouds  11.027412         2   
2012-10-01 16:00:00         0.0   overcast clouds  11.150133         2   
2012-10-01 17:00:00         0.0   overcast clouds  11.272855         2   
2012-10-01 18:00:00         0.0   overcast clouds  11.395576         2   
2012-10-01 19:00:00         0.0   overcast clouds  11.518297         2   
2012-10-01 20:00:00         0.0   overcast clouds  11.641018         2   
2012-10-01 21:00:00         0.0   overcast clouds  11.763739         2   
2012-10-01 22:00:00         0.0   overcast clouds  11.886461         2   
2012-10-01 23:00:00         0.0   overcast clouds  12.009182         2   
2012-10-02 00:00:00         0.0   overcast clouds  12.131903         2   
2012-10-02 01:00:00         0.0   overcast clouds  12.254624         2   
2012-10-02 02:00:00         0.0   overcast clouds  12.377345         2   
2012-10-02 03:00:00         0.0   overcast clouds  12.500067         2   
2012-10-02 04:00:00         0.0   overcast clouds  12.622788         2   
2012-10-02 05:00:00         0.0   overcast clouds  12.745509         2   
2012-10-02 06:00:00         0.0   overcast clouds  12.868230         2   
2012-10-02 07:00:00         0.0   overcast clouds  12.990952         2   
2012-10-02 08:00:00         0.0   overcast clouds  13.113673         2   
2012-10-02 09:00:00         0.0   overcast clouds  13.236394         2   
2012-10-02 10:00:00         0.0   overcast clouds  13.359115         2   
2012-10-02 11:00:00         0.0   overcast clouds  13.481836         2   
2012-10-02 12:00:00         0.0   overcast clouds  13.604558         2   
2012-10-02 13:00:00         0.0   overcast clouds  13.727279         2   
2012-10-02 14:00:00         0.0   overcast clouds  13.850000         2   
2012-10-02 15:00:00         0.0   overcast clouds  16.150000         2   
2012-10-02 16:00:00         0.0      sky is clear  16.700000         3   
2012-10-02 17:00:00         2.0      sky is clear  17.770000         3   
2012-10-02 18:00:00         2.0   overcast clouds  18.100000         2   
2012-10-02 19:00:00         0.0      sky is clear  18.030000         3   
2012-10-02 20:00:00         0.0      sky is clear  17.700000         3   
2012-10-02 21:00:00         0.0      sky is clear  17.400000         3   
2012-10-02 22:00:00         4.0   overcast clouds  16.840000         2   
2012-10-02 23:00:00         3.0   overcast clouds  15.970000         2   
2012-10-03 00:00:00         4.0   overcast clouds  14.930000         2   
2012-10-03 01:00:00         3.0   overcast clouds  14.790000         2   
2012-10-03 02:00:00         2.0   overcast clouds  14.160000         2   
2012-10-03 03:00:00         3.0   overcast clouds  13.940000         2   
2012-10-03 04:00:00         3.0   overcast clouds  13.550000         2   
2012-10-03 05:00:00         2.0   overcast clouds  13.720000         2   
2012-10-03 06:00:00         3.0     broken clouds  13.720000         2   
2012-10-03 07:00:00         3.0   overcast clouds  13.410000         2   
2012-10-03 08:00:00         2.0      sky is clear  13.410000         3   
2012-10-03 09:00:00         4.0   overcast clouds  13.490000         2   
2012-10-03 10:00:00         3.0   overcast clouds  13.410000         2   
2012-10-03 11:00:00         0.0      sky is clear  13.230000         3   
2012-10-03 12:00:00         2.0   overcast clouds  14.160000         2   
2012-10-03 13:00:00         5.0     broken clouds  15.090000         2   
2012-10-03 14:00:00         6.0  scattered clouds  17.430000         2   

                     TempRate  WindRate  Overall   Weather  
datetime                                                    
2012-10-01 13:00:00         2         3        7  

In [1074]:
chiWea.sort_values(by=['Overall'])

Wind Speed      Description    Celcius  DescRate  \
datetime                                                                
2016-04-26 12:00:00         6.0              fog   7.350000         0   
2016-02-27 15:00:00         6.0             haze   0.640000         0   
2015-01-03 16:00:00         3.0    moderate rain   0.855667         0   
2013-05-24 14:00:00         9.0             mist   8.690000         0   
2014-03-19 00:00:00         6.0  very heavy rain   7.930000         0   
2014-03-19 01:00:00         4.0  very heavy rain   7.400000         0   
2015-01-04 02:00:00         3.0       light rain   2.110667         0   
2014-03-19 08:00:00         5.0       light rain   6.540000         0   
2015-01-04 04:00:00         3.0       light rain   2.397000         0   
2014-03-19 09:00:00         4.0       light rain   6.170000         0   
2014-03-19 11:00:00         3.0       light rain   5.950000         0   
2014-03-19 12:00:00         3.0    moderate rain   6.120000         0   
2014-03-19 13:00:00         3.0    moderate rain   6.280000         0   
2015-01-04 05:00:00         3.0       light rain   2.474000         0   
2014-03-19 14:00:00         4.0    moderate rain   6.610000         0   
2014-03-19 15:00:00         3.0       light rain   6.640000         0   
2014-03-19 16:00:00         4.0       light rain   6.290000         0   
2015-01-04 06:00:00         3.0       light rain   2.295500         0   
2014-03-19 17:00:00         5.0       light rain   4.790000         0   
2014-03-19 23:00:00         8.0       light rain   3.070000         0   
2014-03-20 00:00:00         7.0       light rain   2.490000         0   
2014-03-20 01:00:00         7.0       light rain   2.190000         0   
2014-03-20 03:00:00         6.0       light rain   1.570000         0   
2015-01-04 09:00:00         4.0       light rain   1.704000         0   
2015-01-03 14:00:00         4.0       light rain   0.367333         0   
2014-03-20 04:00:00         6.0       light rain   1.300000         0   
2015-01-03 13:00:00         4.0       light rain  -0.027333         0   
2015-01-03 12:00:00         4.0       light rain   0.184000         0   
2016-03-01 05:00:00         7.0       light snow  -0.718576         0   
2017-03-04 09:00:00         5.0             mist  -1.780000         0   
...                         ...              ...        ...       ...   
2014-05-25 20:00:00         0.0       few clouds  27.680000         3   
2014-07-30 01:00:00         2.0       few clouds  21.924000         3   
2017-07-24 04:00:00         2.0     sky is clear  21.170000         3   
2014-07-30 16:00:00         2.0       few clouds  23.490000         3   
2014-08-20 05:00:00         2.0       few clouds  21.597333         3   
2014-08-20 06:00:00         2.0       few clouds  21.647500         3   
2014-08-20 07:00:00         2.0     sky is clear  21.150333         3   
2014-08-20 08:00:00         2.0     sky is clear  21.063667         3   
2014-08-20 09:00:00         2.0     sky is clear  20.796000         3   
2014-08-20 10:00:00         2.0     sky is clear  20.769667         3   
2014-08-20 11:00:00         2.0     sky is clear  20.562333         3   
2013-09-10 03:00:00         1.0       few clouds  29.890000         3   
2014-08-20 12:00:00         2.0     sky is clear  22.037500         3   
2014-05-24 17:00:00         1.0       few clouds  20.190000         3   
2014-05-25 18:00:00         2.0       few clouds  26.680000         3   
2014-05-24 18:00:00         2.0       few clouds  21.610000         3   
2014-05-24 19:00:00         1.0       few clouds  21.890000         3   
2014-08-20 14:00:00         2.0     sky is clear  22.183667         3   
2014-05-24 20:00:00         0.0       few clouds  22.490000         3   
2014-05-24 21:00:00         1.0       few clouds  21.990000         3   
2014-05-24 22:00:00         1.0       few clouds  21.520000         3   
2014-05-24 23:00:00         0.0       few clouds  20.290000         3   
2014-0

We give an overall count of the weather grades. Unsurprsingly, there is more bad weather than either moderate or good. This would be because a lot of the hours would be overnight when the temperature is lower and the sun is in. 

In [1075]:
wea1 = chiWea.groupby('Weather')[('Overall')].count()
wea1.sort_values(ascending=False)

Weather
Bad         25374
Moderate    13160
Good         6718
Name: Overall, dtype: int64

We save this to a new csv, where it will be used for analysis. 

In [1076]:
chiWea.to_csv('../../data/prep/300_ChiWea.csv')